### PR TITLE
Do not delete user from OpenFGA in a DB transaction to avoid issues with minder auth delete

### PR DESCRIPTION
**Pull request by jhrozek**
# Summary

NOTE: This patch touches a bit delicate part of minder. Please review carefully - I was not super acquainted with the code earlier and I wouldn't be suprised if I missed something.

We've been occasionally hitting issues where `minder auth delete` would
delete the user entry but leave the project around. Recovering from this
state then requires database write permissions to manually clean up the
projects.

At least in the case that I reproduced by accident the flow went like
this:
- user issues auth delete
- the transaction starts
- the user projects are returned and contain a non-empty list
- the user entry is deleted from OpenFGA
- the user entry is staged for deletion from the DB as part of the
  transaction
- one of the projects are attempted to be deleted, but the deletion
  fails. This error rolls back the transaction and deleting the user.
- the end-user gets an error message, thinks 'a fluke' and runs auth
  delete again
- the new run starts, calls list project for user from OpenFGA which
  returns an empty list
- the user is deleted from the db and the transaction is commited
- the project are left behind because openFGA returned none for the user
  because as far as openFGA knows the user is gone
- the next auth login can't create the project for the user becasue it already exists

This patch reorganizes the `DeleteUser` flow is such a way that if the
all the database operations are wrapped in a transaction, but the
OpenFGA delete user call is kept outside; the list of projects before the
transaction and deleting the user entry after it.

The call to delete unmanaged projects is changed so that only the
projects that have other subjects than the one being deleted are
deleted.

## Change Type

- [x] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

I reverted the patch that prevents deleting a project in case it still contains
artifacts as a way to easily break project deletion.

Then without the patch, deleting a project caused the projects to be essentially orphaned
and subsequent auth login calls were all failing.

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

**Review by JAORMX — COMMENTED** [review]
lgtm, did you manage to test this locally with something other than unit tests?

**Comment by eleftherias** [comment]
@jhrozek Do you know why this situation happened:
> one of the projects are attempted to be deleted, but the deletion fails. 

Was it a fluke or something intentional that kept the project from being deleted?

**Comment by jhrozek** [comment]
> @jhrozek Do you know why this situation happened:
> 
> > one of the projects are attempted to be deleted, but the deletion fails.
> 
> Was it a fluke or something intentional that kept the project from being deleted?

It was before f4807f133ce9db30f2cc084f0e494d588796ef22 was fixed, so I couldn't delete the project because it had artifacts attached to one of the repos.

**Comment by jhrozek** [comment]
> lgtm, did you manage to test this locally with something other than unit tests?

I reverted the fix for f4807f133ce9db30f2cc084f0e494d588796ef22 and tried to delete the project while it had artifacts attached to the repos. I got consistent error that I can't remove the project, but I didn't end up in this half-state of no FGA assignments and no user entry, but projects left behind.

**Comment by jhrozek** [comment]
> > @jhrozek Do you know why this situation happened:
> > > one of the projects are attempted to be deleted, but the deletion fails.
> > 
> > 
> > Was it a fluke or something intentional that kept the project from being deleted?
> 
> It was before [f4807f1](https://github.com/stacklok/minder/commit/f4807f133ce9db30f2cc084f0e494d588796ef22) was fixed, so I couldn't delete the project because it had artifacts attached to one of the repos.

but your question was valid - if we think that this change is not 100% or too risky, let's not merge it this week.

**Review by JAORMX — COMMENTED** [review]
I still think this is an improvement and it makes sense. I think @rdimitrov said he'd test it, so I'll wait for that.

**Review by rdimitrov — DISMISSED** [review]
tldr: I haven't found any issues 👍 

I've tested it eventually by using the approach mentioned by @jhrozek by reverting the artifacts fix. I've created sub-projects, confirmed that without this fix we end up in a weird state, then enabled the fix and did the same test where I had my main project and the sub-projects and tried deleting my account. Deletion failed because of the reverted artifact fix but it did not broke anything for me.

**Review by JAORMX — DISMISSED** [review]
shipit

File: internal/projects/delete.go
Comment (evankanderson) [inline internal/projects/delete.go:51]:
```suggestion
	if !hasOtherRoleAssignments(as, subject) {
```

File: internal/projects/delete.go
Comment (evankanderson) [inline internal/projects/delete.go:62]:
```suggestion
func hasOtherRoleAssignments(as []*v1.RoleAssignment, subject string) bool {
```

File: internal/controlplane/identity_events.go
Comment (evankanderson) [inline internal/controlplane/identity_events.go:126]: I'm not sure this comment still makes sense -- we should probably update it.

Given that project IDs are UUIDs, it _seems_ reasonable to assert that we clean up the database first because any "dangling" user-to-project assignments in OpenFGA will be effectively harmless, because we should never end up reusing a project ID.  It seems worthwhile including a comment to that effect.

File: internal/controlplane/identity_events.go
Comment (evankanderson) [inline internal/controlplane/identity_events.go:151]: This is subtle -- we'll end up with userDBID being `nil` if neither the `if` nor the the `else if` match.

Can we use `if usr.IdentitySubject != ""` or `if !errors.Is(err, sql.ErrNoRows)` here?

```suggestion
		// We only delete the user if it still exists in the database
		if usr.IdentitySubject != "" {
			l = l.With().Int32("user_id", usr.ID).Logger()
			l.Debug().Msg("deleting user from database")
			if err := qtx.DeleteUser(ctx, usr.ID); err != nil {
				return db.User{}, fmt.Errorf("error deleting user %v", err)
			}
		}
```

File: internal/controlplane/identity_events.go
Comment (evankanderson) [inline internal/controlplane/identity_events.go:158]: Q: Do we want to return an error here after the first Unmanaged project fails?  Would we want to delete the projects before the user?

A: It turns out the order doesn't matter, because it's happening in a transaction.  It would still feel slightly better to me to delete the user after the projects, but the real effect is minimal.

**Review by evankanderson — COMMENTED** [review]
I have #3098 as well, but this looks generally correct, so let's merge it and then I'll merge #3098 after that.

**Review by evankanderson — DISMISSED** [review]

**Comment by evankanderson** [comment]
This fixes the other half of #2948 that I was worried about, more elegantly than I was imagining, because I still forget about database transactions...

**Review by evankanderson — APPROVED** [review]

**Pull request merged**
